### ISSUE TITLE
adapters: Pass artifactProvider to widget adapters and support asynchronous create()

### ISSUE THREAD

**Issue by jpommerening** [LaxarJS/laxar#390]
Currently all widget adapters have a synchronous API. They require lists of controls and widgets to be passed to their `bootstrap()` function where they do some initialization and store the widget modules in a map. Each adapter then implement it's own widget lookup in `create()`.

If we make the `artifactProvider` available to widget adapters they can use the provider's lookup mechanism instead. The plain and React adapters don't really do any bootstrapping at all and could defer widget lookup until the `create()` method is called. (Same with the yet-to-be-finished Vue.js adapter.)

Aside from the additional service, this requires a minor (but backwards compatible) API change for widget adapters in that they need to be allowed to return a Promise in `create()`.

The first argument to `bootstrap()` contains the lists of controls and widgets implemented in the adapter's technology. The only adapter that really needs that argument seems to be the Angular (1) adapter.

**Comment by jpommerening** [comment]
@alex3683 @x1B, about those tests:
I fixed the plain adapter spec test so that it handle the newly asynchronous `adater.create()`. The widget loader spec still creates a synchronous widget adapter mock, so we have both kinds covered (more or less).
Can you have a look and point be to any problems that need to be addressed before we can merge this?

**Comment by x1B** [comment]
LGTM; rebasing and merging. If we find problems, there's still time to mend.

**Comment by x1B** [comment]
@jpommerening lest we forget: and would you mind adding an "Upgrade Information" entry as a comment into this issue?

**Comment by x1B** [comment]
Implemented on `master` (`v2.0.0`).

**Comment by jpommerening** [comment]
## Upgrade Information

Widget adapters are now free to return a Promise in their create method which is then expected to resolve to the widget adapter interface consisting of the `domAttachTo`, `domDetach` and (optionally) `destroy` methods. The `artifactProvider` was added to the services supplied to the widget adapter's `bootstrap` method.

```js
export function bootstrap( { widgets, controls }, { artifactProvider } ) {
   //                                new service -- ^^^^^^^^^^^^^^^^
   return {
      technology,
      create( { widgetName } ) {
         const modulePromise = artifactProvider.forWidget( widgetName )
            .then( provider => provider.module() );

         //     vvvvvvvvvvvvv -- return promise
         return modulePromise
            .then( () => ( {
               domAttachTo() { /* ... */ },
               domDetach() { /* ... */ }
            } ) );
      }
   };
};
```
